Fill the initial hidden state of GRULayer with output from a previous layer

almath123 opened the pull request:
I have added an optional incoming layer `prefill_h` to the GRULayer so that it can be used as part of an encoder-decoder model. It should not be difficult to do this for the other recurrent layers. Also included is a simple example of its usage.

I have been using this approach in my own work and thought it worth feeding back into the project.

Closes #462.

skaae commented:
I think we have a Issue for this: https://github.com/Lasagne/Lasagne/issues/462

It is nearly supported through the `hid_init` argument which needs to be modified to accept input_layers. I think you are more than welcome to solve https://github.com/Lasagne/Lasagne/issues/462 ? @craffel ?

craffel commented:
Yes, better to solve this as part of #462.

f0k commented:
+1 for #462 -- however, you're reasonably close to this. We'd be happy if you could adapt this PR to use the proposed interface in #462 instead. That is, instead of a new argument `prefill_h`, allow `hid_init` to be a layer instance.
The encoder-decoder example is great as well and asked for a lot on the mailing list, but I think this might be suited better for [Lasagne/Recipes](https://github.com/Lasagne/Recipes) just because we're trying to keep the `examples/` directory to a bare minimum.

almath123 commented:
Ok, sure. I have now had a look at #462 . I will use `hid_init` rather than `prefill_h` as the layer input and remove the encoder-decoder example.

f0k commented on lasagne/layers/recurrent.py:
```diff
@@ -803,8 +819,20 @@ def __init__(self, incoming, num_units,
+        if isinstance(cell_init, Layer):
+            incomings.append(cell_init)
+            self.cell_init_incoming_index = len(incomings)-1
+            learn_init = False
```
You shouldn't set `learn_init = False` here -- otherwise providing just one of `hid_init` or `cell_init` as a Layer will disable learning of the other. As it's only used within the `self.add_param()` call further below, the value of `learn_init` doesn't matter when the initial state is provided as a Layer anyway. So for consistence, you can also remove the `learn_init = False` assignment for the other recurrent layer classes.

f0k commented on lasagne/layers/recurrent.py:
```diff
@@ -1174,6 +1218,10 @@ class GRULayer(MergeLayer):
+    prefill_h : :class:`lasagne.layers.Layer
+        Layer which is used to fill the initial hidden state, used for building
+        encoder-decoder models. Default `None`, which means the initial hidden
+        state is not pre-filled from a previous layer.
```
Leftover from previous implementation.

f0k commented on lasagne/layers/recurrent.py:
```diff
@@ -1304,15 +1361,20 @@ def get_output_for(self, inputs, **kwargs):
+            this layer has a mask input (i.e. was instantiated with `mask_input
+            != None`, indicating that the lengths of sequences in each batch
+            vary), `inputs` should have length at least 2, where `inputs[1]` is
+            the `mask`.  The `mask` should be supplied as a Theano variable
```
Please leave it as before, to have a cleaner history (you only changed some line breaks here, no content).

almath123 replied:
What I changed here was to specify "length at least 2" rather than "length 2". I have changed it back for now. Though I think it is a necessary addition for clarity. Thoughts?

f0k commented on lasagne/layers/recurrent.py:
```diff
@@ -1304,15 +1361,20 @@ def get_output_for(self, inputs, **kwargs):
+            Layer)`) `inputs` should have length at least 2. If an input mask
+            was supplied then `inputs[2]` is the hidden state to prefill with.
+            If no input mask was supplied then `inputs[1] is the hidden state
+            to prefill with.
```
A bit shorter: 
```
When the hidden state [...] `inputs` should have length at least 2, and `inputs[-1]` is the hidden state to prefill with.
```

f0k commented on lasagne/tests/layers/test_recurrent.py:
```diff
@@ -59,6 +59,50 @@ def test_recurrent_nparams_learn_init():
+    l_rec = RecurrentLayer(l_inp, 7, hid_init=l_inp_h_de)
+
+    x = T.tensor3()
+    h = T.matrix()
```
`x` and `h` are not used here.

f0k commented on lasagne/tests/layers/test_recurrent.py:
```diff
@@ -59,6 +59,50 @@ def test_recurrent_nparams_learn_init():
+    h = T.matrix()
+
+    # b, W_hid_to_hid and W_in_to_hid + W + b
+    assert len(lasagne.layers.get_all_params(l_rec, trainable=True)) == 5
```
I would first add a test directly checking whether you can see the layers through hid_init:
```
assert lasagne.layers.get_all_layers(l_rec) == [l_inp, l_inp_h, l_inp_h_de, l_rec]
```
Your test only checks whether the parameters of `hid_init` itself are included, not whether you can fully "see through" it to the incoming layers of `hid_init`.

f0k commented on lasagne/tests/layers/test_recurrent.py:
```diff
@@ -438,6 +482,58 @@ def test_lstm_tensor_init():
+    l_lstm = LSTMLayer(l_inp, 7, hid_init=l_inp_h_de, cell_init=l_inp_cell_de)
+
+    x = T.tensor3()
+    h = T.matrix()
```
Same as above: `x` and `h` are unused, and a `get_all_layers()` test would be good! Counting the parameters is not so important then, actually, it's a rather indirect test.

f0k commented on lasagne/tests/layers/test_recurrent.py:
```diff
@@ -704,6 +800,52 @@ def test_gru_init_val_error():
+    l_gru = GRULayer(l_inp, 7, hid_init=l_inp_h_de)
+
+    x = T.tensor3()
+    h = T.matrix()
```
(Same as above.)

f0k commented:
Thank you, looks great except for a few minor comments. Kudos for tackling all three recurrent layers -- I didn't want to ask for it, but it's certainly appreciated! :+1:
In addition to my comments above, the docstrings need to be updated to mention that `hid_init` and `cell_init` can also be `Layer` instances now.

The `self.incoming_mask_index` etc. makes me think that it would be great if the `MergeLayer` also supported dictionaries -- i.e., take a dictionary like `{'input': layer1, `mask`: layer2, `hid_init`: layer3}` instead of a list `[layer1, layer2, layer3]` in the constructor and then provide that dictionary in `get_output_for()`. This would simplify the code a lot. That will need coordination with `lasagne.layers.get_output()` and #337, though.

@skaae: We could get rid of the `hid_init is TensorVariable` special case at the same time -- you don't need it any more, do you? That would remove some `if` clauses and make the docstrings easier. But we can also address this afterwards in a separate PR.

almath123 commented:
Thanks for the comments! I have had a go at integrating them. If there are more fixes needed for this PR I'm happy to make them.

f0k commented:
Thanks, looks good to me! Can you squash your commits into one?
```
git reset --soft 7c0f477
git commit --amend -m 'Have recurrent layers accept `hid_init` and `cell_init` to be a Layer'
git push --force
```

skaae commented:
I missed your commet about tensorvariable. Im fine with removing it. Sorry for being slow

skaae commented:
@almath: it would be great if you could remove the tensorvariable option. The docstring and a few if statements should be modified and the tests should be removed. Do you want to do this? The pr looks good. Thanks!

f0k commented:
> @almath: it would be great if you could remove the tensorvariable option.

This could also be done in a follow-up PR. I wouldn't want to squash this in with the other commits anyway, so whatever you decide to do, please first squash what you have so far!

To not forget about it, when removing the explicit TensorVariable case and adapting the docstring (it's still indirectly supported via `add_param()`, but with changed semantics), it would be good to also document which shapes are expected for `hid_init` and `cell_init` (they differ when it's a Layer). But as I said, this can be done in a follow-up PR.

almath123 commented:
I have squashed my commits. I will leave removing the TensorVariable case to another PR, which i'm happy to do though it could be a couple of days before I get round to it. (If anybody else would prefer to do it; go for it).

craffel commented:
I'll review this tomorrow. Thanks for it!

f0k commented:
I'm fine with merging, but another pair of eyes is welcome :)

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -214,8 +214,14 @@ def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
         # inputs - the layer input, and the mask.  We will just provide the
```
This comment needs updating now.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -356,7 +364,10 @@ def get_output_for(self, inputs, **kwargs):
-            of sequence i)``.
+            of sequence i)``. When the hidden state of this layer is to be
+            pre-filled (i.e. was instantiated with `isinstance(hid_init,
+            Layer)`) `inputs` should have length at least 2, and `inputs[-1]`
```
This should probably say ` ``isinstance(hid_init, Layer) == True`` `.  Note the double backticks.

f0k replied:
I think `== True` can be omitted, it doesn't change that expression's value. Double backticks would be correct, yes. Or maybe reformulate as
```
When the hidden state of this layer is to be
pre-filled (i.e., `hid_init` was set to a
:class:`Layer` instance), ...
```

craffel replied:
Well, it's awkward to just write `isinstance(hid_init, Layer)`, because there's no conditional there, it's just an expression.  It's sort of like saying "was instantiated with ``23*100``".  Huh?  I like your word-explanation better than either.

craffel commented:
> I'm fine with merging, but another pair of eyes is welcome :)

Well, this was something I was supposed to do, so I feel like I should at least look at it!

I made some pedantic comments.  It also feels like `incomings` should be a dict instead of a list, because having the semantics of different objects be defined by their absolute position in a list seems unpythonic to me, but that's just me.  It would also replace `self.mask_incoming_index = -1` and `self.mask_incoming_index > 0` with something much more readable, such as `'mask' in self.incomings`.  Also, I would love to see the tests testing actual expected behavior given some synthetic input values (i.e. compiling a Theano function and passing some dummy input).

f0k commented:
> It also feels like incomings should be a dict instead of a list, because having the semantics of different objects be defined by their absolute position in a list seems unpythonic to me, but that's just me.

No, it's not just you. There are use cases for which a list is appropriate (e.g., the ConcatLayer), and there are others for which a dict is appropriate (e.g., the RecurrentLayer). The easiest solution (in terms of minimal changes needed in Lasagne) would be a second MergeLayer base class that uses dictionaries instead of lists (for `incomings`, `inputs` and `input_shapes`). For the recurrent containers, we'd also like layers to output dictionaries rather than single tensors, though, and we might want to feed this into a MergeLayer as well (i.e., give it a single multiple-output layer rather than multiple single-output layers)... we should think about this. I'll add an issue.

f0k commented:
@almath123: There are two minor comments by @craffel -- would be cool if you could address them so we can merge! Otherwise I can try to do it as part of the merge. Cheers!

almath123 commented:
@f0k sure I will do it later today

almath123 commented:
I have made the suggested changes to the doc strings  and the comments. 

@craffel Thanks for the review and comments. I wasn't sure what needed to be done for further tests. I thought something like this might work (though I haven't pushed it):

```python
def test_recurrent_hid_init_layer_eval():
    # test hid_init as a Layer with some dummy input and init
    in_shp = (2, 2, 3)
    in_h_shp = (2, 5)
    n_units = 7

    # dummy init layers
    W = np.ones((in_h_shp[-1], n_units), dtype=theano.config.floatX)
    W_in_to_hid = np.ones((in_shp[-1], n_units), dtype=theano.config.floatX)
    W_hid_to_hid = np.ones((n_units, n_units), dtype=theano.config.floatX)

    l_inp = InputLayer(in_shp)
    l_inp_h = InputLayer(in_h_shp)
    l_inp_h_de = DenseLayer(l_inp_h, n_units, W=W, nonlinearity=None)
    l_rec = RecurrentLayer(l_inp, n_units, hid_init=l_inp_h_de,
                           W_in_to_hid=W_in_to_hid, W_hid_to_hid=W_hid_to_hid,
                           nonlinearity=None)

    X = T.tensor3()
    Xh = T.matrix()

    output = lasagne.layers.get_output(l_rec, {l_inp: X, l_inp_h: Xh})

    # test net with dummy input
    X_test = np.ones(in_shp, dtype=theano.config.floatX)
    Xh_test = np.ones(in_h_shp, dtype=theano.config.floatX)
    output_val = output.eval({X: X_test, Xh: Xh_test})

    # check output has the right type
    assert isinstance(output_val, np.ndarray)

    # check output is close to the expected in the first iteration
    H_de = np.dot(Xh_test, W)
    res = np.dot(X_test[:, 0, :], W_in_to_hid) + np.dot(H_de, W_hid_to_hid)
    assert np.allclose(output_val[:, 0, :], res)
```

craffel commented:
> @craffel Thanks for the review and comments. I wasn't sure what needed to be done for further tests. I thought something like this might work (though I haven't pushed it):

Looks reasonable to me!

almath123 commented:
@f0k Thanks for picking that up. I have updated the comments I missed and squashed my commits. I haven't added that extra testcase. I was thinking of doing a similar test for each of the recurrent layers and adding it as another PR because I don't want to hold this up any longer.

f0k commented:
Great, let's merge this! Thanks again for your contribution and your patience in addressing our nitpicks.

> I was thinking of doing a similar test for each of the recurrent layers and adding it as another PR

Sounds good. It'd also be good to put the encoder/decoder example somewhere, e.g., on [Lasagne/Recipes](/Lasagne/Recipes). But take your time!
